server/src/instant/util/async.clj

```diff
@@ -118,3 +126,53 @@
               (a/alt!
                 ch ([v] v)
                 timeout-ch :timeout))))))
+
+
+(defn buf-capacity [buf]
+  (cond (instance? FixedBuffer buf)
+        (.n ^FixedBuffer buf)
+
+        (instance? SlidingBuffer buf)
+        (.n ^SlidingBuffer buf)
+
+        (instance? DroppingBuffer buf)
+        (.n ^DroppingBuffer buf)
+
+        (instance? PromiseBuffer buf)
+        1
+
+        :else -1))
+
+(defn gauged-chan
+  "Creates a channel that will publish its size in gauges.
+   chan-name should be a namespaced keyword, e.g. ::my-channel
+
+   There's a small cost to setting it up and it adds 3 attrs per
+   channel, so should be used sparingling with long-running channels."
+  ([chan-name] (gauged-chan chan-name nil))
+  ([chan-name buf-or-n] (gauged-chan chan-name buf-or-n nil))
+  ([chan-name buf-or-n xform] (gauged-chan chan-name buf-or-n xform nil))
+  ([chan-name buf-or-n xform ex-handler]
+   (assert (keyword? chan-name) "chan-name must be a namespaced keyword")
+   (assert (namespace chan-name) "chan-name must be a namespaced keyword")
+   (when xform (assert buf-or-n "buffer must be supplied when transducer is"))
+   (let [buf-or-n (or buf-or-n 1)
+         buf (if (number? buf-or-n) (a/buffer buf-or-n) buf-or-n)
+         chan (a/chan xform ex-handler)]
+     (gauges/add-gauge-metrics-fn
+      (fn [{:keys [cleanup]}]
+        (if (a-impl/closed? chan)
+          (cleanup)
+          [{:path (format "%s.%s.count"
+                          (namespace chan-name)
+                          (name chan-name))
+            :value (count buf)}
```

Comment:
TIL I had no idea you could count the buffer, to get the items that are still in it